File: web/static/native-search.js
export function openNativeSearch({api, targets = []}) {
  const prior = document.activeElement;
  const root = document.createElement('dialog');
  root.className = 'native-history native-search';
  root.setAttribute('aria-label', 'Search saved conversations');
  root.innerHTML = `<header><h2>Search saved conversations</h2><button class="ns-close" aria-label="Close saved conversation search">Close</button></header>
    <section class="ns-browse"><form class="ns-form"><label>Conversation text<input class="ns-query" type="search" required maxlength="500" placeholder="Find something discussed…" autocomplete="off"></label><div class="ns-filters"><label>Target<select class="ns-target" aria-label="Target"><option value="">All targets</option></select></label><label>Agent<select class="ns-agent" aria-label="Agent"><option value="">Claude and Codex</option><option value="claude">Claude</option><option value="codex">Codex</option></select></label><button type="submit" class="ns-submit">Search</button><button type="button" class="ns-stop" hidden>Stop search</button></div></form>
    <p class="ns-status" role="status">Search saved messages across workspaces, including conversations you no longer track.</p>
    <button class="ns-retry" hidden>Retry connection</button><details class="ns-progress" hidden><summary>Target progress</summary><div></div></details>
    <div class="ns-results" aria-label="Saved conversation results"></div>
    <details class="ns-advanced"><summary>Search options</summary><p>If a transcript was rewritten, rebuild its search index. Saved conversations remain unchanged.</p><button class="ns-rebuild">Rebuild and search</button></details></section>
    <section class="ns-reader" hidden><div class="ns-reader-head"><button class="ns-back">Back to results</button><p class="ns-location"></p></div><p class="ns-read-status" role="status"></p><div class="nh-messages"></div></section>`;
  const $ = selector => root.querySelector(selector);
  for (const target of targets) {
    const option = document.createElement('option'); option.value = target.id; option.textContent = target.name;
    $('.ns-target').append(option);
  }
  let closed = false, generation = 0, readGeneration = 0, job = null, timer = null, starting = false;
  let lastResult = null, pollGeneration = 0, selectedResult = null;
  const fit = () => root.style.setProperty('--search-height', `${window.visualViewport?.height || innerHeight}px`);
  fit(); window.visualViewport?.addEventListener('resize', fit);
  const cancel = id => api(`/conversation-search/${id}`, {method:'DELETE'});
  function render(result) {
    lastResult = result;
    $('.ns-stop').hidden = result.done;
    const ready = result.results.length;
    const unfinished = result.scopes.filter(s => s.state === 'queued' || s.state === 'indexing').length;
    const problems = result.scopes.filter(s => s.error || s.progress.issues?.length || s.progress.oversized_entries);
    $('.ns-status').textContent = `${ready} ${ready === 1 ? 'conversation' : 'conversations'} found${unfinished ? ` · searching ${unfinished} ${unfinished === 1 ? 'profile' : 'profiles'}…` : result.complete ? '' : ' · some profiles could not be fully searched'}${result.scopes.some(s => s.more) ? ' · more matches available; narrow your search' : ''}${!ready && result.complete ? '. Try different words or filters.' : ''}`;
    $('.ns-progress').hidden = !result.scopes.length;
    $('.ns-progress summary').textContent = `Target progress · ${result.scopes.length} ${result.scopes.length === 1 ? 'profile' : 'profiles'}${problems.length ? ` · ${problems.length} with issues` : ''}`;
    const progress = $('.ns-progress div'); progress.replaceChildren();
    for (const scope of result.scopes) {
      const p = document.createElement('p');
      p.textContent = `${scope.target} · ${scope.agent}: ${scope.error || scope.state} · ${scope.progress.documents} conversations${scope.progress.pending_files ? ` · ${scope.progress.pending_files} pending` : ''}${scope.progress.oversized_entries ? ` · ${scope.progress.oversized_entries} oversized entries skipped` : ''}${scope.progress.issues?.length ? ' · '+scope.progress.issues.join('; ') : ''}`;
      progress.append(p);
    }
    const list = $('.ns-results'), scroll = list.scrollTop, focused = document.activeElement?.dataset.resultId;
    const existing = new Map([...list.children].map(node => [node.dataset.resultId, node]));
    for (const hit of result.results) {
      let button = existing.get(hit.id);
      if (!button) {
        button = document.createElement('button'); button.type = 'button'; button.className = 'ns-result'; button.dataset.resultId = hit.id;
        button.append(document.createElement('strong'), document.createElement('small'), document.createElement('span'));
      }
      button.children[0].textContent = hit.title;
      button.children[1].textContent = `${hit.target} · ${hit.agent} · ${hit.cwd}`;
      button.children[2].textContent = hit.snippet;
      button.onclick = () => read(hit, result.id);
      list.append(button); existing.delete(hit.id);
    }
    for (const node of existing.values()) node.remove();
    if (focused) [...list.children].find(n => n.dataset.resultId === focused)?.focus({preventScroll:true});
    list.scrollTop = scroll;
  }
  async function poll(id, version) {
    const request = ++pollGeneration;
    try {
      const result = await api(`/conversation-search/${id}`);
      if (closed || version !== generation || request !== pollGeneration) return;
      $('.ns-retry').hidden = true; render(result);
      if (!result.done) timer = setTimeout(() => poll(id, version), 600);
    } catch (error) {
      if (closed || version !== generation || request !== pollGeneration) return;
      $('.ns-status').textContent = `Could not update search: ${error.message}. Available results are retained.`;
      $('.ns-retry').hidden = false;
    }
  }
  async function start(reset = false) {
    if (starting || !$('.ns-form').reportValidity()) return;
    const version = ++generation, previous = job; job = null; starting = true;
    clearTimeout(timer); $('.ns-submit').disabled = true; $('.ns-rebuild').disabled = true;
    $('.ns-stop').hidden = true; $('.ns-retry').hidden = true;
    $('.ns-status').textContent = 'Starting search…';
    try {
      if (previous && !lastResult?.done) { try { await cancel(previous); } catch (error) { if (error.status !== 404) throw error; } }
      if (closed || version !== generation) return;
      const result = await api('/conversation-search', {method:'POST', body:{query:$('.ns-query').value, target_id:Number($('.ns-target').value) || undefined, agent:$('.ns-agent').value || undefined, reset}});
      if (closed || version !== generation) { cancel(result.id).catch(() => {}); return; }
      job = result.id; render(result);
      if (!result.done) timer = setTimeout(() => poll(job, version), 300);
    } catch (error) {
      if (!closed && version === generation) {
        job = previous;
        $('.ns-status').textContent = `Could not start search: ${error.message}`;
        $('.ns-stop').hidden = !job || lastResult?.done;
      }
    } finally {
      starting = false; $('.ns-submit').disabled = false; $('.ns-rebuild').disabled = false;
    }
  }
  async function read(hit, id) {
    const version = ++readGeneration; selectedResult = hit.id;
    $('.ns-browse').hidden = true; $('.ns-reader').hidden = false;
    $('.ns-location').textContent = `${hit.target} · ${hit.agent} · ${hit.cwd}`;
    $('.ns-read-status').textContent = 'Loading matching message…'; $('.nh-messages').replaceChildren(); $('.ns-back').focus();
    try {
      const page = await api(`/conversation-search/${id}/results/${hit.id}`);
      if (closed || version !== readGeneration) return;
      for (const message of page.messages) {
        const card = document.createElement(message.role === 'tool' ? 'details' : 'article');
        card.className = `nh-message nh-${message.role}`;
        if (message.matched) { card.classList.add('ns-match'); if (message.role === 'tool') card.open = true; }
        const heading = document.createElement(message.role === 'tool' ? 'summary' : 'h3');
        heading.textContent = `${message.role === 'user' ? 'You' : message.role === 'tool' ? 'Tool activity' : 'Assistant'}${message.matched ? ' · Matching message' : ''}`;
        const text = document.createElement('pre'); text.textContent = message.text; card.append(heading, text);
        if (message.truncated) { const note = document.createElement('small'); note.textContent = 'Long message shortened in this view.'; card.append(note); }
        $('.nh-messages').append(card);
      }
      $('.ns-read-status').textContent = `Matching message with nearby context${page.changed_neighbors ? ` · ${page.changed_neighbors} changed messages omitted` : ''}${!page.index_complete ? ' · indexing is incomplete' : ''}.`;
      $('.ns-match')?.scrollIntoView({block:'center'});
    } catch (error) {
      if (!closed && version === readGeneration) $('.ns-read-status').textContent = `${error.message}. Return to results and search again.`;
    }
  }
  $('.ns-form').onsubmit = event => { event.preventDefault(); start(); };
  $('.ns-rebuild').onclick = () => start(true);
  $('.ns-retry').onclick = () => { if (job) { clearTimeout(timer); poll(job, generation); } };
  $('.ns-stop').onclick = async () => {
    if (!job) return; const id = job, version = generation;
    try { const result = await cancel(id); if (!closed && version === generation) { clearTimeout(timer); pollGeneration++; render(result); if (!result.done) poll(id, version); } }
    catch (error) { if (!closed && version === generation) $('.ns-status').textContent = `Could not stop search: ${error.message}`; }
  };
  $('.ns-back').onclick = () => { readGeneration++; $('.ns-reader').hidden = true; $('.ns-browse').hidden = false; [...$('.ns-results').children].find(n => n.dataset.resultId === selectedResult)?.focus({preventScroll:true}); };
  root.addEventListener('keydown', event => {
    event.stopPropagation();
    if (event.isComposing) return;
    const buttons = [...$('.ns-results').children];
    if ((event.key === 'ArrowDown' || event.key === 'ArrowUp') && (event.target === $('.ns-query') || buttons.includes(event.target))) {
      event.preventDefault(); const index = buttons.indexOf(event.target);
      buttons[Math.max(0, Math.min(buttons.length-1, index+(event.key === 'ArrowDown' ? 1 : -1)))]?.focus();
    }
  });
  $('.ns-close').onclick = () => root.close();
  root.addEventListener('close', () => {
    closed = true; generation++; readGeneration++; clearTimeout(timer);
    if (job && !lastResult?.done) cancel(job).catch(() => {});
    window.visualViewport?.removeEventListener('resize', fit); root.remove(); if (prior?.isConnected) prior.focus();
  });
  document.body.append(root); root.showModal(); $('.ns-query').focus(); return root;
}

File: internal/console/native_search.go
package console

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/charmbracelet/bubbles/viewport"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/x/ansi"
)

type nativeSearchHit struct {
	ID, Title, Target, Agent, Cwd, Snippet string
}
type nativeSearchScope struct {
	Target, Agent, State, Error string
	More                        bool
	Progress                    struct {
		Documents int
		Pending   int `json:"pending_files"`
		Oversized int `json:"oversized_entries"`
		Issues    []string
	}
}
type nativeSearchData struct {
	ID             string
	Done, Complete bool
	Results        []nativeSearchHit
	Scopes         []nativeSearchScope
}
type nativeSearchState struct {
	data                                        nativeSearchData
	query, agent, target, failure, mode, reader string
	selected, readGeneration, generation        int
	starting                                    bool
	viewport                                    viewport.Model
}
type nativeSearchMsg struct {
	owner      *nativeSearchState
	generation int
	started    bool
	data       nativeSearchData
	err        error
}
type nativeSearchTick struct {
	owner      *nativeSearchState
	generation int
}
type nativeSearchReadMsg struct {
	owner      *nativeSearchState
	generation int
	data       []byte
	err        error
}

func (m *dashboard) nativeSearchForm() tea.Cmd {
	targets := []choice{{"All targets", ""}}
	for _, r := range m.targets {
		targets = append(targets, choice{name(r), id(r)})
	}
	return m.openForm("Search saved conversations", []field{
		{Key: "query", Label: "Conversation text", Required: true},
		{Key: "target", Label: "Target", Options: targets},
		{Key: "agent", Label: "Agent", Options: []choice{{"Claude and Codex", ""}, {"Claude", "claude"}, {"Codex", "codex"}}},
	}, func(values map[string]any) tea.Cmd {
		query := strings.TrimSpace(str(values["query"]))
		if len(query) > 500 {
			m.notice = "Search query must be at most 500 characters"
			return nil
		}
		m.form = nil
		s := &nativeSearchState{query: query, target: str(values["target"]), agent: str(values["agent"]), viewport: viewport.New(max(1, m.width-4), max(1, m.height-7))}
		m.nativeSearch = s
		return m.startNativeSearch(false)
	})
}
func (m *dashboard) startNativeSearch(reset bool) tea.Cmd {
	s := m.nativeSearch
	if s == nil || s.starting {
		return nil
	}
	s.starting = true
	s.failure = ""
	s.generation++
	generation := s.generation
	c := m.client
	previous := s.data.ID
	body := map[string]any{"query": s.query, "agent": s.agent, "reset": reset}
	if target, err := strconv.ParseInt(s.target, 10, 64); err == nil {
		body["target_id"] = target
	}
	return func() tea.Msg {
		if previous != "" {
			if _, err := c.JSON("DELETE", "/conversation-search/"+url.PathEscape(previous), nil); err != nil {
				var status *HTTPError
				if !errors.As(err, &status) || status.Status != 404 {
					return nativeSearchMsg{owner: s, generation: generation, err: err}
				}
			}
		}
		b, err := c.JSON("POST", "/conversation-search", body)
		var data nativeSearchData
		if err == nil {
			err = json.Unmarshal(b, &data)
		}
		return nativeSearchMsg{owner: s, generation: generation, started: true, data: data, err: err}
	}
}
func (m *dashboard) cancelNativeSearch(s *nativeSearchState) tea.Cmd {
	c := *m.client
	h := http.Client{}
	if c.HTTP != nil {
		h = *c.HTTP
	}
	h.Timeout = 3 * time.Second
	c.HTTP = &h
	id := s.data.ID
	if id == "" {
		return nil
	}
	s.generation++
	generation := s.generation
	return func() tea.Msg {
		b, err := c.JSON("DELETE", "/conversation-search/"+url.PathEscape(id), nil)
		var data nativeSearchData
		if err == nil {
			err = json.Unmarshal(b, &data)
		}
		return nativeSearchMsg{owner: s, generation: generation, data: data, err: err}
	}
}
func (m *dashboard) pollNativeSearch(s *nativeSearchState) tea.Cmd {
	c := m.client
	id := s.data.ID
	generation := s.generation
	return func() tea.Msg {
		b, err := c.JSON("GET", "/conversation-search/"+url.PathEscape(id), nil)
		var data nativeSearchData
		if err == nil {
			err = json.Unmarshal(b, &data)
		}
		return nativeSearchMsg{owner: s, generation: generation, data: data, err: err}
	}
}
func (m *dashboard) receiveNativeSearch(v nativeSearchMsg) tea.Cmd {
	s := m.nativeSearch
	if s != v.owner || s.generation != v.generation {
		if v.started && v.err == nil && v.data.ID != "" && !v.data.Done {
			return m.cancelNativeSearch(&nativeSearchState{data: v.data})
		}
		return nil
	}
	s.starting = false
	if v.err != nil {
		s.failure = clean(v.err.Error())
		return nil
	}
	previousOffset := s.viewport.YOffset
	preserveOffset := len(s.data.Results) > 0
	selected := ""
	if s.selected < len(s.data.Results) {
		selected = s.data.Results[s.selected].ID
	}
	s.data = v.data
	s.failure = ""
	s.selected = min(s.selected, max(0, len(s.data.Results)-1))
	for i, hit := range s.data.Results {
		if hit.ID == selected {
			s.selected = i
			break
		}
	}
	m.renderNativeSearch()
	if preserveOffset {
		s.viewport.SetYOffset(previousOffset)
	}
	if !s.data.Done {
		generation := s.generation
		return tea.Tick(700*time.Millisecond, func(time.Time) tea.Msg { return nativeSearchTick{s, generation} })
	}
	return nil
}
func (m *dashboard) readNativeSearch() tea.Cmd {
	s := m.nativeSearch
	if len(s.data.Results) == 0 {
		return nil
	}
	hit := s.data.Results[s.selected]
	s.mode = "reader"
	s.reader = "Loading matching message…"
	s.readGeneration++
	generation := s.readGeneration
	m.renderNativeSearch()
	c := m.client
	path := "/conversation-search/" + url.PathEscape(s.data.ID) + "/results/" + url.PathEscape(hit.ID)
	return func() tea.Msg { b, err := c.JSON("GET", path, nil); return nativeSearchReadMsg{s, generation, b, err} }
}
func (m *dashboard) receiveNativeSearchRead(v nativeSearchReadMsg) {
	s := m.nativeSearch
	if s != v.owner || s.readGeneration != v.generation || s.mode != "reader" {
		return
	}
	if v.err != nil {
		s.reader = "Could not read match: " + clean(v.err.Error()) + "\nEsc returns to results; r retries the search."
	} else {
		var page struct {
			Messages []struct {
				Role, Text         string
				Matched, Truncated bool
			}
			Changed  int  `json:"changed_neighbors"`
			Complete bool `json:"index_complete"`
		}
		if err := json.Unmarshal(v.data, &page); err != nil {
			s.reader = "Could not parse matching context"
		} else {
			var lines []string
			for _, message := range page.Messages {
				title := strings.ToUpper(message.Role)
				if message.Matched {
					title += " · MATCHING MESSAGE"
				}
				text := title + "\n" + message.Text
				if message.Truncated {
					text += "\n[Long message shortened]"
				}
				lines = append(lines, text)
			}
			if page.Changed > 0 {
				lines = append(lines, fmt.Sprintf("%d changed neighboring messages omitted", page.Changed))
			}
			if !page.Complete {
				lines = append(lines, "Indexing is incomplete; this is the available context")
			}
			s.reader = strings.Join(lines, "\n\n")
		}
	}
	m.renderNativeSearch()
	s.viewport.GotoTop()
	// Center the matching heading even when the earlier context is long.
	for i, line := range strings.Split(ansi.Wrap(clean(s.reader), max(1, m.width-4), ""), "\n") {
		if strings.Contains(line, "· MATCHING MESSAGE") {
			s.viewport.SetYOffset(max(0, i-2))
			break
		}
	}
}
func (m *dashboard) renderNativeSearch() {
	s := m.nativeSearch
	if s == nil {
		return
	}
	s.viewport.Width = max(1, m.width-4)
	s.viewport.Height = max(1, m.height-4-len(strings.Split(ansi.Wrap(nativeSearchHelp, max(1, m.width-2), ""), "\n")))
	if s.mode == "reader" {
		s.viewport.SetContent(ansi.Wrap(clean(s.reader), s.viewport.Width, ""))
		return
	}
	var lines []string
	selectedStart := 0
	if s.mode == "progress" {
		for _, scope := range s.data.Scopes {
			lines = append(lines, ansi.Wrap(clean(fmt.Sprintf("%s · %s · %s\n%d conversations · %d pending · %d oversized\n%s %s", scope.Target, scope.Agent, scope.State, scope.Progress.Documents, scope.Progress.Pending, scope.Progress.Oversized, scope.Error, strings.Join(scope.Progress.Issues, "; "))), s.viewport.Width, ""))
		}
	} else {
		for i, hit := range s.data.Results {
			prefix := "  "
			if i == s.selected {
				prefix = "› "
				selectedStart = len(strings.Split(strings.Join(lines, "\n\n"), "\n"))
				if len(lines) > 0 {
					selectedStart++
				} else {
					selectedStart = 0
				}
			}
			snippet := []rune(clean(hit.Snippet))
			if len(snippet) > 300 {
				snippet = append(snippet[:300], '…')
			}
			text := ansi.Wrap(clean(prefix+hit.Title+"\n"+hit.Target+" · "+hit.Agent+" · "+hit.Cwd+"\n"+string(snippet)), s.viewport.Width, "")
			if i == s.selected {
				text = accent.Render(text)
			}
			lines = append(lines, text)
		}
	}
	if len(lines) == 0 {
		lines = append(lines, "No matches yet. Search includes saved histories outside tracked workspaces.")
	}
	s.viewport.SetContent(strings.Join(lines, "\n\n"))
	if s.mode == "" && (selectedStart < s.viewport.YOffset || selectedStart >= s.viewport.YOffset+s.viewport.Height-2) {
		s.viewport.SetYOffset(selectedStart)
	}
}
func (m *dashboard) updateNativeSearch(k tea.KeyMsg) tea.Cmd {
	s := m.nativeSearch
	switch k.String() {
	case "ctrl+c", "q":
		return tea.Sequence(m.cancelNativeSearch(s), tea.Quit)
	case "esc":
		if s.mode != "" {
			s.mode = ""
			s.readGeneration++
			m.renderNativeSearch()
			return nil
		}
		m.nativeSearch = nil
		return m.cancelNativeSearch(s)
	case "n":
		if s.starting {
			return nil
		}
		m.nativeSearch = nil
		return tea.Batch(m.cancelNativeSearch(s), m.nativeSearchForm())
	case "r":
		if s.starting {
			return nil
		}
		s.mode = ""
		return m.startNativeSearch(false)
	case "R":
		if s.starting {
			return nil
		}
		s.mode = ""
		return m.startNativeSearch(true)
	case "s":
		if s.starting {
			return nil
		}
		return m.cancelNativeSearch(s)
	case "p":
		if s.mode == "progress" {
			s.mode = ""
		} else {
			s.mode = "progress"
		}
		m.renderNativeSearch()
		s.viewport.GotoTop()
		return nil
	case "enter":
		if s.mode == "" && !s.starting {
			return m.readNativeSearch()
		}
	case "up", "k":
		if s.mode == "" {
			s.selected = max(0, s.selected-1)
			m.renderNativeSearch()
			return nil
		}
	case "down", "j":
		if s.mode == "" {
			s.selected = min(max(0, len(s.data.Results)-1), s.selected+1)
			m.renderNativeSearch()
			return nil
		}
	}
	var cmd tea.Cmd
	s.viewport, cmd = s.viewport.Update(k)
	return cmd
}

const nativeSearchHelp = "↑↓ choose/scroll · Enter read · p progress · s stop · r retry · R rebuild · n new · Esc back · q quit"

func (m *dashboard) nativeSearchView() string {
	if m.width < 35 || m.height < 12 {
		return "Search saved conversations\nResize to at least 35 × 12.\nEsc back · q quit"
	}
	s := m.nativeSearch
	state := "Searching…"
	if s.data.Done {
		state = "Search complete"
		if !s.data.Complete {
			state = "Search finished with incomplete profiles; p shows details"
		}
	}
	if s.starting {
		state = "Starting search…"
	}
	if s.failure != "" {
		state = s.failure + " · r retries"
	}
	noun := "conversations"
	if len(s.data.Results) == 1 {
		noun = "conversation"
	}
	summary := fmt.Sprintf("%d %s · %s", len(s.data.Results), noun, state)
	for _, scope := range s.data.Scopes {
		if scope.More {
			summary += " · narrow query for more matches"
			break
		}
	}
	return accent.Bold(true).Render(" Search saved conversations") + "\n" + ansi.Truncate(oneLine(s.query), max(1, m.width-2), "…") + "\n" + ansi.Truncate(oneLine(summary), max(1, m.width-2), "…") + "\n\n" + s.viewport.View() + "\n" + ansi.Wrap(nativeSearchHelp, max(1, m.width-2), "")
}
